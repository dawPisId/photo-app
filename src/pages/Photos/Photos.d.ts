interface idInterface {
  id: string;
}

interface PhotosProps {
  match: string;
}
