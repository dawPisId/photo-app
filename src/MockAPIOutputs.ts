export const mockAlbumsData = [
  {
    userId: 1,
    id: 1,
    title: "quidem molestiae enim",
  },
  {
    userId: 1,
    id: 2,
    title: "sunt qui excepturi placeat culpa",
  },
  {
    userId: 1,
    id: 3,
    title: "omnis laborum odio",
  },
  {
    userId: 1,
    id: 4,
    title: "non esse culpa molestiae omnis sed optio",
  },
  {
    userId: 1,
    id: 5,
    title: "eaque aut omnis a",
  },
  {
    userId: 1,
    id: 6,
    title: "natus impedit quibusdam illo est",
  },
  {
    userId: 1,
    id: 7,
    title: "quibusdam autem aliquid et et quia",
  },
  {
    userId: 1,
    id: 8,
    title: "qui fuga est a eum",
  },
  {
    userId: 1,
    id: 9,
    title: "saepe unde necessitatibus rem",
  },
  {
    userId: 1,
    id: 10,
    title: "distinctio laborum qui",
  },
  {
    userId: 2,
    id: 11,
    title: "quam nostrum impedit mollitia quod et dolor",
  },
  {
    userId: 2,
    id: 12,
    title: "consequatur autem doloribus natus consectetur",
  },
  {
    userId: 2,
    id: 13,
    title: "ab rerum non rerum consequatur ut ea unde",
  },
  {
    userId: 2,
    id: 14,
    title: "ducimus molestias eos animi atque nihil",
  },
  {
    userId: 2,
    id: 15,
    title: "ut pariatur rerum ipsum natus repellendus praesentium",
  },
  {
    userId: 2,
    id: 16,
    title: "voluptatem aut maxime inventore autem magnam atque repellat",
  },
  {
    userId: 2,
    id: 17,
    title: "aut minima voluptatem ut velit",
  },
  {
    userId: 2,
    id: 18,
    title: "nesciunt quia et doloremque",
  },
  {
    userId: 2,
    id: 19,
    title: "velit pariatur quaerat similique libero omnis quia",
  },
  {
    userId: 2,
    id: 20,
    title: "voluptas rerum iure ut enim",
  },
];

export const mockAlbumCardData = [
  {
    albumId: 1,
    id: 1,
    title: "accusamus beatae ad facilis cum similique qui sunt",
    url: "https://via.placeholder.com/600/92c952",
    thumbnailUrl: "https://via.placeholder.com/150/92c952",
  },
  {
    albumId: 1,
    id: 2,
    title: "reprehenderit est deserunt velit ipsam",
    url: "https://via.placeholder.com/600/771796",
    thumbnailUrl: "https://via.placeholder.com/150/771796",
  },
  {
    albumId: 1,
    id: 3,
    title: "officia porro iure quia iusto qui ipsa ut modi",
    url: "https://via.placeholder.com/600/24f355",
    thumbnailUrl: "https://via.placeholder.com/150/24f355",
  },
  {
    albumId: 1,
    id: 4,
    title: "culpa odio esse rerum omnis laboriosam voluptate repudiandae",
    url: "https://via.placeholder.com/600/d32776",
    thumbnailUrl: "https://via.placeholder.com/150/d32776",
  },
];

export const mockZoomData = [
  {
    albumId: 1,
    id: 10,
    title: "beatae et provident et ut vel",
    url: "https://via.placeholder.com/600/810b14",
    thumbnailUrl: "https://via.placeholder.com/150/810b14",
  },
];

export const mockPhotosData = [
  {
    albumId: 1,
    id: 1,
    title: "accusamus beatae ad facilis cum similique qui sunt",
    url: "https://via.placeholder.com/600/92c952",
    thumbnailUrl: "https://via.placeholder.com/150/92c952",
  },
  {
    albumId: 1,
    id: 2,
    title: "reprehenderit est deserunt velit ipsam",
    url: "https://via.placeholder.com/600/771796",
    thumbnailUrl: "https://via.placeholder.com/150/771796",
  },
  {
    albumId: 1,
    id: 3,
    title: "officia porro iure quia iusto qui ipsa ut modi",
    url: "https://via.placeholder.com/600/24f355",
    thumbnailUrl: "https://via.placeholder.com/150/24f355",
  },
  {
    albumId: 1,
    id: 4,
    title: "culpa odio esse rerum omnis laboriosam voluptate repudiandae",
    url: "https://via.placeholder.com/600/d32776",
    thumbnailUrl: "https://via.placeholder.com/150/d32776",
  },
  {
    albumId: 1,
    id: 5,
    title: "natus nisi omnis corporis facere molestiae rerum in",
    url: "https://via.placeholder.com/600/f66b97",
    thumbnailUrl: "https://via.placeholder.com/150/f66b97",
  },
  {
    albumId: 1,
    id: 6,
    title: "accusamus ea aliquid et amet sequi nemo",
    url: "https://via.placeholder.com/600/56a8c2",
    thumbnailUrl: "https://via.placeholder.com/150/56a8c2",
  },
  {
    albumId: 1,
    id: 7,
    title: "officia delectus consequatur vero aut veniam explicabo molestias",
    url: "https://via.placeholder.com/600/b0f7cc",
    thumbnailUrl: "https://via.placeholder.com/150/b0f7cc",
  },
  {
    albumId: 1,
    id: 8,
    title: "aut porro officiis laborum odit ea laudantium corporis",
    url: "https://via.placeholder.com/600/54176f",
    thumbnailUrl: "https://via.placeholder.com/150/54176f",
  },
  {
    albumId: 1,
    id: 9,
    title: "qui eius qui autem sed",
    url: "https://via.placeholder.com/600/51aa97",
    thumbnailUrl: "https://via.placeholder.com/150/51aa97",
  },
  {
    albumId: 1,
    id: 10,
    title: "beatae et provident et ut vel",
    url: "https://via.placeholder.com/600/810b14",
    thumbnailUrl: "https://via.placeholder.com/150/810b14",
  },
  {
    albumId: 1,
    id: 11,
    title: "nihil at amet non hic quia qui",
    url: "https://via.placeholder.com/600/1ee8a4",
    thumbnailUrl: "https://via.placeholder.com/150/1ee8a4",
  },
  {
    albumId: 1,
    id: 12,
    title:
      "mollitia soluta ut rerum eos aliquam consequatur perspiciatis maiores",
    url: "https://via.placeholder.com/600/66b7d2",
    thumbnailUrl: "https://via.placeholder.com/150/66b7d2",
  },
  {
    albumId: 1,
    id: 13,
    title: "repudiandae iusto deleniti rerum",
    url: "https://via.placeholder.com/600/197d29",
    thumbnailUrl: "https://via.placeholder.com/150/197d29",
  },
  {
    albumId: 1,
    id: 14,
    title: "est necessitatibus architecto ut laborum",
    url: "https://via.placeholder.com/600/61a65",
    thumbnailUrl: "https://via.placeholder.com/150/61a65",
  },
  {
    albumId: 1,
    id: 15,
    title: "harum dicta similique quis dolore earum ex qui",
    url: "https://via.placeholder.com/600/f9cee5",
    thumbnailUrl: "https://via.placeholder.com/150/f9cee5",
  },
  {
    albumId: 1,
    id: 16,
    title: "iusto sunt nobis quasi veritatis quas expedita voluptatum deserunt",
    url: "https://via.placeholder.com/600/fdf73e",
    thumbnailUrl: "https://via.placeholder.com/150/fdf73e",
  },
  {
    albumId: 1,
    id: 17,
    title: "natus doloribus necessitatibus ipsa",
    url: "https://via.placeholder.com/600/9c184f",
    thumbnailUrl: "https://via.placeholder.com/150/9c184f",
  },
  {
    albumId: 1,
    id: 18,
    title: "laboriosam odit nam necessitatibus et illum dolores reiciendis",
    url: "https://via.placeholder.com/600/1fe46f",
    thumbnailUrl: "https://via.placeholder.com/150/1fe46f",
  },
  {
    albumId: 1,
    id: 19,
    title: "perferendis nesciunt eveniet et optio a",
    url: "https://via.placeholder.com/600/56acb2",
    thumbnailUrl: "https://via.placeholder.com/150/56acb2",
  },
  {
    albumId: 1,
    id: 20,
    title:
      "assumenda voluptatem laboriosam enim consequatur veniam placeat reiciendis error",
    url: "https://via.placeholder.com/600/8985dc",
    thumbnailUrl: "https://via.placeholder.com/150/8985dc",
  },
];
