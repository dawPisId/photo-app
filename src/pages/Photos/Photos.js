import React, { useEffect, useState } from "react";

import Card from "../../components/Card/Card";
import Grid from "../../components/Grid/Grid";
import TopSearchBar from "../../components/TopSearchBar/TopSearchBar";

function Photos() {
  const [photos, setPhotos] = useState([]);
  useEffect(() => {
    fetch("https://jsonplaceholder.typicode.com/albums/2/photos")
      .then((response) => response.json())
      .then((responseData) => setPhotos(responseData))
      .catch((error) => {
        console.log("Error", error);
      });
  }, []);

  return (
    <div>
      <TopSearchBar label="Your Photos" />
      <Grid>
        {photos.map((photo) => (
          <Card key={photo.id} name={photo.title} url={photo.thumbnailUrl} />
        ))}
      </Grid>
    </div>
  );
}

export default Photos;
