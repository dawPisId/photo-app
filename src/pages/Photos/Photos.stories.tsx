import { ComponentMeta, ComponentStory } from "@storybook/react";

import { BrowserRouter } from "react-router-dom";
import Photos from "./Photos";
import React from "react";

export default {
  title: "Photo App/Pages/Photos",
  component: Photos,
} as ComponentMeta<typeof Photos>;

const Template: ComponentStory<typeof Photos> = (args) => (
  <BrowserRouter>
    <Photos />
  </BrowserRouter>
);

export const Main = Template.bind({});
Main.args = {};
