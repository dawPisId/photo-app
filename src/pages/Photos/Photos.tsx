import React, { FC, useEffect, useState } from "react";

import Card from "../../components/Card/Card";
import Grid from "../../components/Grid/Grid";
import ImageZoom from "../../components/ImageZoom/ImageZoom";
import Loader from "../../components/Loader/Loader";
import PageSelect from "../../components/PageSelect/PageSelect";
import TopSearchBar from "../../components/TopSearchBar/TopSearchBar";
import { URL } from "../../APIAddress";
import { idInterface } from "./Photos.d";
import { useParams } from "react-router-dom";
import { zoom } from "../../components/PageSelect/helpers";

/**
 * Component used to display the final layout of the Photos page.
 */
const Photos: FC = () => {
  const [query, setQuery] = useState("");
  const [photos, setPhotos] = useState([]);
  const [page, setPage] = useState(1);
  const [pageCount, setPageCount] = useState(1);
  const [itemsPerPage, setItemsPerPage] = useState(12);
  const [isLoading, setIsLoading] = useState(true);
  const [isZoomed, setIsZoomed] = useState(false);
  const [zoomUrl, setZoomUrl] = useState("");
  const [zoomLabel, setZoomLabel] = useState("");
  const { id }: idInterface = useParams();

  useEffect(() => {
    const valueCheck = typeof id === "string";
    const path = valueCheck ? `albums/${id}/photos` : "photos";
    const firstPhotoIndex = (page - 1) * itemsPerPage;
    setIsLoading(true);
    fetch(`${URL}/${path}?q=${query}`)
      .then((response) => response.json())
      .then((responseData) => {
        setPageCount(Math.ceil(responseData.length / itemsPerPage));
        setPhotos(
          responseData.slice(firstPhotoIndex, firstPhotoIndex + itemsPerPage)
        );
      })
      .catch((error) => {
        console.log("error", error);
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, [itemsPerPage, id, page, query]);

  return (
    <div>
      {isZoomed && (
        <ImageZoom
          data-testid="testZoom"
          setZoomTrigger={setIsZoomed}
          imgUrl={zoomUrl}
          imgLabel={zoomLabel}
        />
      )}
      <TopSearchBar label="Your Photos" setSearchQuery={setQuery} />
      {isLoading ? (
        <Loader />
      ) : (
        <div data-testid="testPhoto">
          <Grid>
            {photos.map(({ id, title, thumbnailUrl, url }) => (
              <Card
                key={id}
                name={title}
                url={thumbnailUrl}
                action={() => {
                  zoom(setIsZoomed, setZoomUrl, url, setZoomLabel, title);
                }}
              />
            ))}
          </Grid>
        </div>
      )}
      <PageSelect
        currentPage={page}
        currentPageSetter={setPage}
        pagesCount={pageCount}
        itemCount={itemsPerPage}
        itemsPerPageSetter={setItemsPerPage}
      />
    </div>
  );
};

export default Photos;
