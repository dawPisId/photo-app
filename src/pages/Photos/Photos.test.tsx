import "@testing-library/jest-dom/extend-expect";

import { BrowserRouter, Route } from "react-router-dom";
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { mockPhotosData, mockZoomData } from '../../MockAPIOutputs';

import Photos from "./Photos";
import React from "react";

it("displays the loader when loading", () => {
  act(() => {
    const testRender = render(
      <BrowserRouter>
        <Route exact path="/" component={Photos} />
      </BrowserRouter>
    );
    expect(testRender).toMatchSnapshot();
  });
});

it("renders correctly", async () => {
  const mockFetch = Promise.resolve({
    json: () => Promise.resolve(mockPhotosData),
  });
  jest.spyOn(window, "fetch").mockImplementationOnce(() => mockFetch as any);
  await act(async () => {
    const testRender = render(
      <BrowserRouter>
        <Route exact path="/" component={Photos} />
      </BrowserRouter>
    );
    await waitFor(() => testRender.getByTestId("testPhoto"));
    expect(testRender).toMatchSnapshot();
  });
});

it("displays zoomed image on card click", async () => {
  const mockFetch = Promise.resolve({
    json: () => Promise.resolve(mockZoomData),
  });
  jest.spyOn(window, "fetch").mockImplementationOnce(() => mockFetch as any);
  await act(async () => {
    const testRender = render(
      <BrowserRouter>
        <Route exact path="/" component={Photos} />
      </BrowserRouter>
    );
    await waitFor(() => testRender.getByTestId("testCard"));
    fireEvent.click(screen.getByTestId("testCard"));
    await waitFor(() => testRender.getByTestId("testZoom"));
    expect(testRender).toMatchSnapshot();
  });
});

it("changes per page count", async () => {
  const mockFetch = Promise.resolve({
    json: () => Promise.resolve(mockPhotosData),
  });
  jest.spyOn(window, "fetch").mockImplementation(() => mockFetch as any);
  await act(async () => {
    const testRender = render(
      <BrowserRouter>
        <Route exact path="/" component={Photos} />
      </BrowserRouter>
    );
    await waitFor(() => testRender.getAllByTestId("testCard"));
    fireEvent.click(screen.getByTestId("testList"));
    await waitFor(() => testRender.getAllByTestId("testListItem"));
    fireEvent.click(screen.getAllByTestId("testListItem")[1]);
    testRender.rerender(
      <BrowserRouter>
        <Route exact path="/" component={Photos} />
      </BrowserRouter>
    );
    await waitFor(() => testRender.getAllByTestId("testCard"));
    expect(testRender.getAllByTestId("testCard").length).toEqual(16);
  });
});
