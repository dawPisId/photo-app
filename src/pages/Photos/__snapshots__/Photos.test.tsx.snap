// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="sc-bkbkJK duhwYw"
  >
    <span
      className="sc-iemWCZ iyOWzq"
    >
      Your Photos
    </span>
    <div
      className="sc-hHEiqL jWQJxi"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <span
        className="sc-kfYoZR fIMaiM"
      >
        Search by title
      </span>
      <input
        className="sc-dlMDgC fEVeCl"
        onChange={[Function]}
        type="text"
      />
    </div>
    <button
      className="sc-dIvrsQ jxlbCO"
      onClick={[Function]}
    >
      SEARCH
    </button>
  </div>
  <div
    className="sc-pNWdM dHCxGu"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root sc-iCoGMd dZSfOz"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 8l-4 4h3c0 3.31-2.69 6-6 6-1.01 0-1.97-.25-2.8-.7l-1.46 1.46C8.97 19.54 10.43 20 12 20c4.42 0 8-3.58 8-8h3l-4-4zM6 12c0-3.31 2.69-6 6-6 1.01 0 1.97.25 2.8.7l1.46-1.46C15.03 4.46 13.57 4 12 4c-4.42 0-8 3.58-8 8H1l4 4 4-4H6z"
      />
    </svg>
    <div
      className="sc-fujyAs ezswet"
    >
      Loading
    </div>
  </div>
  <div>
    <div
      className="sc-jrsJWt cypHCg"
    >
      <div
        className="sc-fnVZcZ dRnWYw"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
        <div
          className="sc-fFSPTT fLhJfU active"
          onClick={[Function]}
        >
          1
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
          />
        </svg>
      </div>
      <div
        className="sc-iqAclL jtBzoG"
      >
        <div
          className="sc-crzoAE lkwUSO"
          onClick={[Function]}
        >
          <span
            className="sc-hBMUJo hMsLWB"
          >
            Per page
          </span>
          <div
            className="sc-ksluID gdOUob"
          >
            12
          </div>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <div
          className="sc-dIsUp leREAs hidden"
        >
          <div
            className="sc-bqGGPW jZazZi"
            onClick={[Function]}
          >
            12
          </div>
          <div
            className="sc-bqGGPW jZazZi"
            onClick={[Function]}
          >
            16
          </div>
          <div
            className="sc-bqGGPW jZazZi"
            onClick={[Function]}
          >
            20
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
